utils/tail/tail.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package tail implements "tail -F" functionality following rotated logs
+package tail
+
+import (
+	"bufio"
+	"io"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+)
+
+type Tail struct {
+	reader   *bufio.Reader
+	filename string
+	file     *os.File
+	stop     chan bool
+}
+
+const pollInterval = time.Second
+
+// NewTail starts opens the given file and watches it for deletion/rotation
+func NewTail(filename string) (t *Tail, err error) {
+	t = &Tail{
+		filename: filename,
+	}
+	t.file, err = os.Open(t.filename)
+	if err != nil {
+		glog.Errorf("Open failed on %s", t.filename)
+		return nil, err
+	}
+	t.file.Seek(0, os.SEEK_END)
+	t.reader = bufio.NewReader(t.file)
+	t.stop = make(chan bool)
+	go t.watchFile()
+	return t, nil
+}
+
+// Read implements the io.Reader interface for Tail
+func (t *Tail) Read(p []byte) (n int, err error) {
+	return t.reader.Read(p)
+}
+
+// type assertion
+var _ io.Reader = &Tail{}
+
+// Close stops watching and closes the file
+func (t *Tail) Close() {
+	t.file.Close()
+	close(t.stop)
+}
+
+const maxReopenAttempts = 3
+
+func (t *Tail) watchFile() {
+	for {
+		origfs, err := os.Stat(t.filename)
+		if err != nil {
+			glog.Errorf("Stat failed on %s", t.filename)
+			return
+		}
+		for {
+			time.Sleep(pollInterval)
+			select {
+			case <-time.After(pollInterval):
+			case <-t.stop:
+				return
+			}
+			fs, err := os.Stat(t.filename)
+			if err != nil || !os.SameFile(origfs, fs) {
+				glog.Infof("File %s deleted/moved", t.filename)
+				break
+			}
+		}
+		t.file.Close()
+		attempt := 1
+		for attempt <= maxReopenAttempts {
```

Comment:
Use a `for atempt := 0; attempt < maxReopenAttempts; attempt++` style loop (I don't think you're incrementing attempt)